src/collision.cpp

```diff
@@ -453,7 +449,14 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 
 		if (nearest_collided == COLLISION_AXIS_NONE) {
 			// No collision with any collision box.
-			*pos_f += truncate(*speed_f * dtime, 100.0f);
+			*pos_f += truncate(aspeed_f * dtime, 100.0f);
+			// Final speed:
+			*speed_f += accel_f * dtime;
+			// Limit speed for avoiding hangs
+			speed_f->Y = rangelim(speed_f->Y, -5000, 5000);
+			speed_f->X = rangelim(speed_f->X, -5000, 5000);
+			speed_f->Z = rangelim(speed_f->Z, -5000, 5000);
+			*speed_f = truncate(*speed_f, 10000.0f);
```

Comment:
This code is repeated 3 times. Would you please be so nice to move this to a separate function?